### PR TITLE
mattermost: 4.8.0 -> 4.9.1

### PR DESCRIPTION
**Pull request by ryantm**
They changed the location of the main command, and it looks like you can build it directly from the top-level, so that simplifies things. Also, they changed how versions are displayed, so we don't need to pass the version number anymore. Unfortunately, it looks like I had to add back in the thing setting the interpreter of the output binary. I haven't had the time to track down why that is necessary again, maybe @mic92 would know.

- [x] built on NixOS
- [x] tested NixOS service

### CODE REVIEW

**Comment by ryantm** [comment]
Darwin doesn't build because it wants patchelf, I wonder if the interpreter was working correctly on darwin when it used to build successfully https://github.com/NixOS/nixpkgs/pull/37200#issuecomment-373862503

File: pkgs/servers/mattermost/default.nix [hunk 1]
```diff
@@ -13,30 +12,23 @@ buildGoPackage rec {
...
-    mkdir -p $bin
+  postInstall = ''
     tar --strip 1 -C $bin -xf $webApp
+    ln -s $bin/bin/platform $bin/bin/mattermost-platform
   '';
 
-  postInstall = ''
-    ln -s $bin/bin/platform $bin/bin/mattermost-platform
+  postFixup = ''
+    patchelf --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) $bin/bin/platform
```
Comment (Mic92) [inline pkgs/servers/mattermost/default.nix]: Why is this necessary? 

```console
$ /nix/store/6b0jynb0szqj81cxsl8y1ivhswxw3wn9-mattermost-4.9.1-bin/bin/mattermost-platform  --help
```

works for me

```
$ ldd /nix/store/6b0jynb0szqj81cxsl8y1ivhswxw3wn9-mattermost-4.9.1-bin/bin/mattermost-platform
        linux-vdso.so.1 (0x00007f866dfe6000)
        libpthread.so.0 => /nix/store/27x7pinqdsl9f3rpbm8bsszd9fhwq266-glibc-2.27/lib/libpthread.so.0 (0x00007f866dba5000)
        libc.so.6 => /nix/store/27x7pinqdsl9f3rpbm8bsszd9fhwq266-glibc-2.27/lib/libc.so.6 (0x00007f866d7f1000)
        /lib64/ld-linux-x86-64.so.2 => /nix/store/27x7pinqdsl9f3rpbm8bsszd9fhwq266-glibc-2.27/lib64/ld-linux-x86-64.so.2 (0x00007f866ddc4000)
```

Comment (ryantm) [reply pkgs/servers/mattermost/default.nix]: I'm glad it works for you, but it doesn't for me:

```
$ ldd /nix/store/ph0pc99dkpsiz72lwcz173icylzvphkg-mattermost-4.9.1-bin/bin/mattermost-platform
        linux-vdso.so.1 =>  (0x00007fff81928000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f6c5f3fd000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f6c5f033000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f6c5f61a000)
```
I assume you are building it on NixOS? I'm building it on Ubuntu 16.04 with nix-build version 1.11.13.

**Comment by ryantm** [comment]
@Mic92 Removing the postFixup phase did not work for on NixOS either:

```
$ nix-build --version
nix-build (Nix) 2.0
$ nix-build -A mattermost
/nix/store/6b0jynb0szqj81cxsl8y1ivhswxw3wn9-mattermost-4.9.1-bin
$ file /nix/store/6b0jynb0szqj81cxsl8y1ivhswxw3wn9-mattermost-4.9.1-bin/bin/platform
/nix/store/6b0jynb0szqj81cxsl8y1ivhswxw3wn9-mattermost-4.9.1-bin/bin/platform: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, not stripped
$ nix-info
system: "x86_64-linux", multi-user?: yes, version: nix-env (Nix) 2.0, channels(root): "", channels(ryantm): "nixos-17.03-17.03.902.536d3cac2e", nixpkgs: /home/ryantm/p/nixpkgs
```
ldd fails to run on the binary with "No such file or directory".

With the postFixup phase:

```
$ nix-build -A mattermost
/nix/store/cjfhbsd9ff814637x66zciawvssdsa6x-mattermost-4.9.1-bin
[ryantm@home1:~/p/nixpkgs]$ file /nix/store/cjfhbsd9ff814637x66zciawvssdsa6x-mattermost-4.9.1-bin/bin/platform
/nix/store/cjfhbsd9ff814637x66zciawvssdsa6x-mattermost-4.9.1-bin/bin/platform: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /nix/store/27x7pinqdsl9f3rpbm8bsszd9fhwq266-glibc-2.27/lib/ld-linux-x86-64.so.2, not stripped
[ryantm@home1:~/p/nixpkgs]$ ldd /nix/store/cjfhbsd9ff814637x66zciawvssdsa6x-mattermost-4.9.1-bin/bin/platform
        linux-vdso.so.1 (0x00007fff189ea000)
        libpthread.so.0 => /nix/store/27x7pinqdsl9f3rpbm8bsszd9fhwq266-glibc-2.27/lib/libpthread.so.0 (0x00007f60a420b000)
        libc.so.6 => /nix/store/27x7pinqdsl9f3rpbm8bsszd9fhwq266-glibc-2.27/lib/libc.so.6 (0x00007f60a3e57000)
        /nix/store/27x7pinqdsl9f3rpbm8bsszd9fhwq266-glibc-2.27/lib/ld-linux-x86-64.so.2 (0x00007f60a442a000)
```

**Comment by ryantm** [comment]
I tried deleting the postFixup-removed instantiation and rebuilding it with `--option sandbox true` and it was still messed up.

**Comment by Mic92** [comment]
I had for reasons a symlink in /lib64 for the glibc link loader. That's why I was confused it worked on my machine. The problem is that the `platform` command is not inside the go project that we are building from source. Instead it is coming from the tarball. Do you where the source for this tarball is? It might be better to build the go binary there, otherwise we have to restrict the platform to x86_64.

**Comment by ryantm** [comment]
Ah, so I thought my changes were building the platform from the src, but it must have just been getting it from the webApp instead. I'll close this and work on fixing it.
